src/middlewared/middlewared/plugins/nfs_/krb5.py

```diff
@@ -0,0 +1,95 @@
+import errno
+from middlewared.service import private, Service
+from middlewared.service_exception import CallError
+from middlewared.plugins.idmap import DSType
+from middlewared.plugins.directoryservices import DSStatus
+from middlewared.schema import accepts, Dict, Str
+
+
+class NFSService(Service):
+
+    class Config:
+        service = "nfs"
+        service_verb = "restart"
+        datastore_prefix = "nfs_srv_"
+        datastore_extend = 'nfs.nfs_extend'
+
+    @private
+    async def validate_directoryservices(self, ds):
+        """
+        Validation currently only succeeds in case of HEALTHY AD
+        directory service. At future point, depending on demand, the
+        same convenience feature can be added for the LDAP directory
+        service.
+        """
+        ds = await self.middleware.call('directoryservices.get_state')
+        ad_status = DSStatus[ds['activedirectory']]
+        ldap_status = DSStatus[ds['ldap']]
+
+        if ldap_status != DSStatus.DISABLED:
+            raise CallError('Automated addition of NFS SPN records has '
+                            'not yet been implemented for the LDAP directory'
+                            'service.', errno=errno.EINVAL)
+
+        if ad_status != DSStatus.HEALTHY:
+            raise CallError('Active Directory Directory Service is currently'
+                            f'[{ds["activedirectory"]}]. Status must be HEALTHY '
+                            'in order to successfully add a kerberos SPN entry.')
+
+    @private
+    async def add_principal_ad(self, data):
+        """
+        Typically elevated permissions are required to make SPN changes.
+        Pass user-provided credentials to our kinit method rather than
+        relying on the existing kerberos ticket / principal.
+        """
+        ad = await self.middleware.call('activedirectory.config')
+        ad['dstype'] = DSType.DS_TYPE_ACTIVEDIRECTORY.value
+        ad['bindname'] = data.get("username", "")
+        ad['bindpw'] = data.get("password", "")
+        ad['kerberos_principal'] = ''
+
+        await self.middleware.call('kerberos.do_kinit', ad)
+        return await self.middleware.call('activedirectory.add_nfs_spn', ad)
```

Comment:
What happens if the SPN entry already exists on an object in AD? Will this return appropriately or at least hint to the end-user why it failed to add?

Reply:
Just to document this... if we fail to set the SPN in AD, then we raise a CallError with information about why it failed. Whether the information contained in stderr from setspn will be of use to support is up for debate, but at least they will know that an error of sorts occurred.